### ISSUE TITLE
GELF Log Driver not retry sending if it failed only one times

### ISSUE THREAD

**Issue by winggundamth** [docker/docker#16330]
We're using Graylog with by listen GELF udp protocol and have all our containers sending log to Graylog with logging driver parameter. Everything doing fine until we upgrade Graylog today that need to down Graylog host for 5 minutes and when it went back. The log from every containers not send to Graylog anymore.

To reproduce

1. Configure Docker container to use GELF logging driver to send to Graylog
2. Stop Graylog for few minutes and start Graylog again
3. You will see log is not sending to Graylog anymore until you restart container 

What I have to do to get all containers sending log to Graylog again is restart every containers that won't be good since many containers run in production.

I'm check upstart Docker log and found this only one line

```
^[[31mERRO^[[0m[55795] Failed to log msg "    2015/09/16 01:36:16 [INFO] agent: Synced check
'service:chat-preproduction-nodejs_5000'" for logger gelf: gelf: cannot send GELF message: write udp 172.24.12.5:12201: connection refused
```

```
ubuntu@chatcafe-pre-service-01:~$ docker version
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:19:00 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:19:00 UTC 2015
 OS/Arch:      linux/amd64

ubuntu@chatcafe-pre-service-01:~$ docker info
Containers: 5
Images: 92
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: gelf
Kernel Version: 4.0.9-040009-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 2
Total Memory: 3.858 GiB
Name: chatcafe-pre-service-01
ID: OVOJ:XKYU:HRD3:2MBN:46LF:UMJS:6BII:ZJEB:UVO2:XFZB:6DNZ:4RBJ
WARNING: No swap limit support
Labels:
 provider=generic

ubuntu@chatcafe-pre-service-01:~$ uname -a
Linux chatcafe-pre-service-01 4.0.9-040009-generic #201507212131 SMP Wed Jul 22 01:33:03 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Is this a bug? Or anyone has work around on this without restarting container?

**Comment by runcom** [comment]
ping @mariussturm is there anything to configure to tell the logger to reconnect if host went down?

**Comment by mariussturm** [comment]
Since the driver is using udp it should simply fire logs over to Graylog, no matter if the server is up or down. I will investigate if I can reproduce this.

**Comment by winggundamth** [comment]
At first I think the same. But after I upgrade my Graylog2 allinone Docker container from 1.1.6 to 1.2.0 by stop the old one and start new one with --volumes-from and same id and keys. Almost all of my containers won't send log to Graylog anymore and I think they should retry later. But until now almost 12 hours has past and those containers still not sending anything to Graylog.

**Comment by phemmer** [comment]
I believe I've seen stuff like this before. It happens when you start a new container listening on a port used by a previous container. The iptables forwarding rules get confused, and connections which existed for the old container (and yes, though UDP is stateless, iptables still treats it as a "connection"), don't get routed to the new container.

I believe there's a way to reset the kernel so that it'll drop the connection state and start routing to the new graylog container. But it's been a while since I've dealt with this issue. Will see about setting up a test later.

**Comment by mariussturm** [comment]
After doing some tests I can say this is definitely not a sender problem, the GELF driver has no problem with a restarted Graylog server. This problem must be on the receiver side. Changed container IP or abandoned iptable states.

**Comment by winggundamth** [comment]
So can we consider this as a Docker bug then?

I hope I could find the way to reset the routing stuffs so we can have workaround for this problem.

**Comment by phemmer** [comment]
Well, your workaround can be to start the (graylog) container with `--net=host`.
Though I'm fairly sure there's a way to reset the connection state (so there would be a second workaround). But that requires a extra manual step when you kill the container.

**Comment by winggundamth** [comment]
From @mariussturm investigation so my workaround is easier than I expected. Just restart Graylog host (Graylog is just only container on that host) and it solved.

But this still need to consider as a bug to permanent fix the problem.

**Comment by mariussturm** [comment]
@winggundamth could you please provide detailed information about your environment, like OS version, network setup, network devices in between... Only if we are able to reproduce the bug we can say in which part of the stack the problem lies.

**Comment by winggundamth** [comment]
Below is information on our boxes. It's set up on AWS EC2 with VPC. Every containers is on the same VPC same network. What I did is doing upgrade Graylog from 1.1.6 to 1.2.0. This is the steps

1. Stop the current running Graylog 1.1.6 container
2. Create new Graylog container from Graylog 1.2.0 Docker Image with --volumes-from Graylog 1.1.6 container with the same exposed ports, same id and same secret.
3. You will see the problem that log is not showing in Graylog

```
ubuntu@chatcafe-pre-service-01:~$ docker version
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:19:00 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:19:00 UTC 2015
 OS/Arch:      linux/amd64

ubuntu@chatcafe-pre-service-01:~$ docker info
Containers: 5
Images: 92
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: gelf
Kernel Version: 4.0.9-040009-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 2
Total Memory: 3.858 GiB
Name: chatcafe-pre-service-01
ID: OVOJ:XKYU:HRD3:2MBN:46LF:UMJS:6BII:ZJEB:UVO2:XFZB:6DNZ:4RBJ
WARNING: No swap limit support
Labels:
 provider=generic

ubuntu@chatcafe-pre-service-01:~$ uname -a
Linux chatcafe-pre-service-01 4.0.9-040009-generic #201507212131 SMP Wed Jul 22 01:33:03 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by cpuguy83** [comment]
Any outcome here?

**Comment by mariussturm** [comment]
Sorry can not reproduce. Works as intended for me. I would close the ticket for now. If there are more people with the same problem please ping me.

**Comment by thaJeztah** [comment]
@winggundamth are you okay with closing this?

**Comment by winggundamth** [comment]
Yes

**Comment by chiefy** [comment]
I actually just ran into this in our VPC.

Stay with me here, because I know next to nothing about UDP. We have a cluster of apps running 14.04.3 LTS w/ docker-compose 1.5.0. Our containers are configured to use the GELF log driver to point to an internal DNS, for this case we'll call it `internal.logstash.mycompany.org`

Everything w/ our ELK stack has been running fine for days, but unfortunately Amazon e-mailed us and told us they had to kill our Logstash instance. I fired a new one up, changed the DNS to point to new private IP, and started tailing the logstash log to make sure logs were getting forwarded.

Nothing.

I ssh'd into one of our app hosts, and noticed the following:
```
ubuntu@ip-10-10-10-1:~$ netstat | grep udp
udp        0      0 10.10.10.1:35293    10.10.10.2:12201    ESTABLISHED
udp        0      0 10.10.10.1:52176    10.10.10.2:12201    ESTABLISHED
udp        0      0 10.10.10.1:40769    10.10.10.2:12201    ESTABLISHED
udp        0      0 10.10.10.1:50376    10.10.10.2:12201    ESTABLISHED
```
10.10.10.2 is the IP of the old logstash instance that went down. It doesn't appear to be a DNS caching issue either, when I dig or ping `internal.logstash.mycompany.org` it resolves to the new IP.

This is going to be a huge deal for us because the only solution I've found is restarting the docker daemon on each host and we have 50+ instances.
